modules/redis/types.go

```diff
@@ -0,0 +1,422 @@
+package redis
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	"github.com/zmap/zgrab2"
+)
+
+var (
+	// ErrInvalidData is returned when the server returns data that cannot be
+	// interpreted as a valid Redis value.
+	ErrInvalidData = errors.New("invalid data")
+
+	// ErrWrongType is returned when one identifier is expected but another is
+	// found.
+	ErrWrongType = errors.New("wrong type specifier")
+
+	// ErrBadLength is returned when an invalid length value is found (e.g. a
+	// negative length or longer than expected).
+	ErrBadLength = errors.New("bad length")
+)
+
+// RedisType is a human readable type identifier for redis data
+type RedisType string
+
+const (
+	// TypeSimpleString identifies Simple String (string) values
+	TypeSimpleString RedisType = "simple string"
+
+	// TypeError identifiers Error (string) values
+	TypeError = "error"
+
+	// TypeInteger identifiers Integer (int64) values
+	TypeInteger = "integer"
+
+	// TypeBulkString identifies Bulk String ([]byte) values
+	TypeBulkString = "bulk string"
+
+	// TypeArray identifies Array ([]RedisValue) types
+	TypeArray = "array"
+)
+
+// RedisValue is implemented by any redis that can be returned by the server
+type RedisValue interface {
+	Type() RedisType
+	Encode() []byte
+}
+
+// SimpleString type -- must not contain \r or \n. https://redis.io/topics/protocol#resp-simple-strings
+type SimpleString string
+
+// Type identifies this instance as a TypeSimpleString
+func (SimpleString) Type() RedisType {
+	return TypeSimpleString
+}
+
+// Encode returns the SimpleString encoding of the value ("+<string value>\r\n").
+func (str SimpleString) Encode() []byte {
+	return []byte("+" + str + "\r\n")
+}
+
+// ErrorMessage type -- a string, where the first word can optionally be
+// interpreted as an error identifier.
+// See https://redis.io/topics/protocol#resp-errors
+type ErrorMessage string
+
+// Type identifies this instance as a TypeError
+func (ErrorMessage) Type() RedisType {
+	return TypeError
+}
+
+// Encode returns the encoding of the error message ("-<error message>\r\n")
+func (err ErrorMessage) Encode() []byte {
+	return []byte("-" + err + "\r\n")
+}
+
+// ErrorPrefix returns the first word of the error message, which can be
+// interpreted as a sort of error code.
+func (err ErrorMessage) ErrorPrefix() string {
+	return strings.SplitN(string(err), " ", 2)[0]
```

Comment:
Do we need a `len(err) > 0` check here?

Reply:
Confirmed that `len(strings.SplitN("", " ", 2)) == 1`, but it would probably be good to make it explicit.